src/EventListener/InitializeSystemListener.php

```diff
@@ -0,0 +1,257 @@
+                    break;
+                }
+            }
+
+            unset($langs, $lang);
+        }
+
+        $GLOBALS['TL_LANGUAGE'] = $_SESSION['TL_LANGUAGE'];
+
+        // Fully load the configuration
+        $objConfig = Config::getInstance();
+
+        // Generate the symlinks before any potential output
+        if (!$objConfig->isComplete() && !is_link(TL_ROOT . '/system/themes/flexible')) {
+            $automator = new Automator();
+            $automator->generateSymlinks();
+        }
+
+        // Show the "insecure document root" message
+        if ('cli' !== PHP_SAPI && 'contao/install.php' !== TL_SCRIPT /* FIXME */ && '/web' !== substr(Environment::get('path'), -4) && !Config::get('ignoreInsecureRoot')) {
+            die_nicely('be_insecure', 'Your installation is not secure. Please set the document root to the <code>/web</code> subfolder.');
+        }
+
+        // Show the "incomplete installation" message
+        if ('cli' !== PHP_SAPI && 'contao/install.php' !== TL_SCRIPT /* FIXME */ && !$objConfig->isComplete()) {
+            die_nicely('be_incomplete', 'The installation has not been completed. Open the Contao install tool to continue.');
+        }
+
+        Input::initialize();
+
+        // Always show error messages if logged into the install tool (see #5001)
+        if (Input::cookie('TL_INSTALL_AUTH') && !empty($_SESSION['TL_INSTALL_AUTH']) && Input::cookie('TL_INSTALL_AUTH') == $_SESSION['TL_INSTALL_AUTH'] && $_SESSION['TL_INSTALL_EXPIRE'] > time()) {
+            Config::set('displayErrors', 1);
+        }
+
+        // Configure the error handling
+        @ini_set('display_errors', (Config::get('displayErrors') ? 1 : 0));
+        error_reporting((Config::get('displayErrors') || Config::get('logErrors')) ? Config::get('errorReporting') : 0);
+
+        // Set the timezone
+        @ini_set('date.timezone', Config::get('timeZone'));
+        @date_default_timezone_set(Config::get('timeZone'));
+
+        // Set the mbstring encoding
+        if (USE_MBSTRING && function_exists('mb_regex_encoding')) {
+            mb_regex_encoding(Config::get('characterSet'));
+        }
+
+        // HOOK: add custom logic (see #5665)
+        if (isset($GLOBALS['TL_HOOKS']['initializeSystem']) && is_array($GLOBALS['TL_HOOKS']['initializeSystem'])) {
+            foreach ($GLOBALS['TL_HOOKS']['initializeSystem'] as $callback) {
+                System::importStatic($callback[0])->$callback[1]();
+            }
+        }
+
+        // Include the custom initialization file
+        if (file_exists(TL_ROOT . '/system/config/initconfig.php')) {
+            include TL_ROOT . '/system/config/initconfig.php';
+        }
+
+        RequestToken::initialize();
+
+        // Check the request token upon POST requests
+        if ($_POST && !RequestToken::validate(Input::post('REQUEST_TOKEN'))) {
+
+            // Force a JavaScript redirect upon Ajax requests (IE requires absolute link)
```

Comment:
Even though this is legacy code I think we should get rid of the `exit;` and either throw a `ResponseException` or a `DieNicelyException`, both upcoming with @discordier's changes. So we could add a FIXME here?